indexer/src/main.rs

```diff
@@ -0,0 +1,186 @@
+use clap::Parser;
+
+pub use cached::SizedCache;
+use futures::future::try_join_all;
+use futures::{try_join, StreamExt};
+use tokio::sync::Mutex;
+use tracing::{debug, info};
+
+use explorer_database::{adapters, models, receipts_cache};
+
+use crate::configs::Opts;
+
+mod configs;
+mod metrics;
+
+// Categories for logging
+const INDEXER_FOR_EXPLORER: &str = "indexer_for_explorer";
+
+async fn handle_message(
+    pool: &explorer_database::actix_diesel::Database<explorer_database::diesel::PgConnection>,
+    streamer_message: near_lake_framework::near_indexer_primitives::StreamerMessage,
+    strict_mode: bool,
+    receipts_cache: receipts_cache::ReceiptsCache,
+) -> anyhow::Result<()> {
+    metrics::BLOCK_COUNT.inc();
+    metrics::LATEST_BLOCK_HEIGHT.set(streamer_message.block.header.height.try_into().unwrap());
+
+    debug!(
+        target: INDEXER_FOR_EXPLORER,
+        "ReceiptsCache #{} \n {:#?}", streamer_message.block.header.height, &receipts_cache
+    );
+    adapters::blocks::store_block(pool, &streamer_message.block).await?;
+
+    // Chunks
+    adapters::chunks::store_chunks(
+        pool,
+        &streamer_message.shards,
+        &streamer_message.block.header.hash,
+    )
+    .await?;
+
+    // Transactions
+    let transactions_future = adapters::transactions::store_transactions(
+        pool,
+        &streamer_message.shards,
+        &streamer_message.block.header.hash,
+        streamer_message.block.header.timestamp,
+        streamer_message.block.header.height,
+        std::sync::Arc::clone(&receipts_cache),
+    );
+
+    // Receipts
+    let receipts_future = adapters::receipts::store_receipts(
+        pool,
+        &streamer_message.shards,
+        &streamer_message.block.header.hash,
+        streamer_message.block.header.timestamp,
+        strict_mode,
+        std::sync::Arc::clone(&receipts_cache),
+    );
+
+    // We can process transactions and receipts in parallel
+    // because most of receipts depend on transactions from previous blocks,
+    // so we can save up some time here.
+    // In case of local receipts (they are stored in the same block with corresponding transaction),
+    // we hope retry logic will cover it fine
```

Comment:
The approach described in this comment works. Though it constantly adds 1 retry and warning in the logs thus making logs unnecessarily noisy. I would vote to separate these processes as I don't think we are winning here much time. @telezhnaya what do you think?

Reply:
Can we benchmark these 2 options and compare the results?
It could be done in the separate PR for sure

Reply:
Anyway, I suggest moving the logic as close to the master state as possible and work on any other ideas separately

Reply:
@telezhnaya I fully agree about separate PR, let's get back to it later. I believe I have created an issue about it a long time ago anyway